File: src/vector_space.cpp
#include "vector_space.h"

#include <optional>
#include <string_view>

#include "absl/strings/numbers.h"
#include "absl/strings/str_format.h"
#include "macros.h"
#include "re2/re2.h"
#include "util.h"

namespace vectorlite {

std::optional<DistanceType> ParseDistanceType(std::string_view distance_type) {
  if (distance_type == "l2") {
    return DistanceType::L2;
  } else if (distance_type == "ip") {
    return DistanceType::InnerProduct;
  } else if (distance_type == "cosine") {
    return DistanceType::Cosine;
  }
  return std::nullopt;
}

std::optional<VectorType> ParseVectorType(std::string_view vector_type) {
  if (vector_type == "float32") {
    return VectorType::Float32;
  }
  return std::nullopt;
}

absl::StatusOr<VectorSpace> VectorSpace::Create(size_t dim,
                                                DistanceType distance_type,
                                                VectorType vector_type) {
  if (dim == 0) {
    return absl::InvalidArgumentError("Dimension must be greater than 0");
  }

  VectorSpace result;
  result.distance_type = distance_type;
  result.normalize = distance_type == DistanceType::Cosine;
  result.vector_type = vector_type;
  switch (distance_type) {
    case DistanceType::L2:
      result.space = std::make_unique<hnswlib::L2Space>(dim);
      break;
    case DistanceType::InnerProduct:
      result.space = std::make_unique<hnswlib::InnerProductSpace>(dim);
      break;
    case DistanceType::Cosine:
      result.space = std::make_unique<hnswlib::InnerProductSpace>(dim);
      break;
    default:
      std::string err_msg =
          absl::StrFormat("Invalid space type: %d", distance_type);
      return absl::InvalidArgumentError(err_msg);
  }

  return result;
}

absl::StatusOr<NamedVectorSpace> CreateNamedVectorSpace(
    size_t dim, DistanceType distance_type, std::string_view vector_name,
    VectorType vector_type) {
  auto result = VectorSpace::Create(dim, distance_type, vector_type);

  if (!result.ok()) {
    return result.status();
  }

  NamedVectorSpace named_vector_space(std::move(*result));
  named_vector_space.vector_name = vector_name;
  return named_vector_space;
}

absl::StatusOr<NamedVectorSpace> NamedVectorSpace::FromString(
    std::string_view space_str) {
  static const re2::RE2 reg(
      "^(?<vector_name>\\w+)\\s+(?<vector_type>\\w+)\\[(?<dim>\\d+)\\]\\s*(?<"
      "distance_type>\\w+)?\\s*$");
  VECTORLITE_ASSERT(reg.ok());

  std::string_view vector_name;
  std::string_view vector_type_str;
  size_t dim = 0;
  std::optional<std::string_view> distance_type_str;
  if (re2::RE2::FullMatch(space_str, reg, &vector_name, &vector_type_str, &dim,
                          &distance_type_str)) {
    if (!IsValidColumnName(vector_name)) {
      std::string error =
          absl::StrFormat("Invalid vector name: %s", vector_name);
      return absl::InvalidArgumentError(error);
    }

    auto vector_type = ParseVectorType(vector_type_str);
    if (!vector_type) {
      std::string error =
          absl::StrFormat("Invalid vector type: %s", vector_type_str);
      return absl::InvalidArgumentError(error);
    }

    DistanceType distance_type = DistanceType::L2;
    if (distance_type_str) {
      auto maybe_distance_type = ParseDistanceType(*distance_type_str);
      if (!maybe_distance_type) {
        std::string error =
            absl::StrFormat("Invalid distance type: %s", *distance_type_str);
        return absl::InvalidArgumentError(error);
      }
      distance_type = *maybe_distance_type;
    }

    return CreateNamedVectorSpace(dim, distance_type, vector_name,
                                  *vector_type);
  }
  return absl::InvalidArgumentError("Unable to parse vector space");
}

size_t VectorSpace::dimension() const {
  return *reinterpret_cast<size_t*>(space->get_dist_func_param());
}

}  // end namespace vectorlite

File: rust/src/vector_space.rs
//! Parsing of the vector-space declaration string, e.g.
//! `my_embedding float32[384] cosine`. Mirrors `vector_space.cpp` /
//! `NamedVectorSpace::FromString` and `util.cpp::IsValidColumnName`.

/// Distance metric. Discriminants must match `VlDistanceType` in core_shim.h.
#[derive(Clone, Copy, PartialEq, Eq, Debug)]
#[repr(i32)]
pub enum DistanceType {
    L2 = 0,
    InnerProduct = 1,
    Cosine = 2,
}

/// Stored element type. Discriminants must match `VlVectorType` in core_shim.h.
#[derive(Clone, Copy, PartialEq, Eq, Debug)]
#[repr(i32)]
pub enum VectorType {
    Float32 = 0,
    BFloat16 = 1,
    Float16 = 2,
}

impl VectorType {
    pub fn element_size(self) -> usize {
        match self {
            VectorType::Float32 => 4,
            VectorType::BFloat16 => 2,
            VectorType::Float16 => 2,
        }
    }
}

pub fn parse_distance_type(s: &str) -> Option<DistanceType> {
    match s {
        "l2" => Some(DistanceType::L2),
        "ip" => Some(DistanceType::InnerProduct),
        "cosine" => Some(DistanceType::Cosine),
        _ => None,
    }
}

pub fn parse_vector_type(s: &str) -> Option<VectorType> {
    match s {
        "float32" => Some(VectorType::Float32),
        "bfloat16" => Some(VectorType::BFloat16),
        "float16" => Some(VectorType::Float16),
        _ => None,
    }
}

#[derive(Clone, Debug)]
pub struct NamedVectorSpace {
    pub vector_name: String,
    pub dim: usize,
    pub distance_type: DistanceType,
    pub vector_type: VectorType,
}

fn is_word_char(c: char) -> bool {
    c.is_ascii_alphanumeric() || c == '_'
}

/// Validates a SQLite column name: must start with a letter or underscore and
/// contain only letters, digits, underscores or `$`. (Keyword rejection from
/// the C++ version is omitted; it is not exercised by the test suite.)
pub fn is_valid_column_name(name: &str) -> bool {
    let mut chars = name.chars();
    match chars.next() {
        Some(c) if c.is_ascii_alphabetic() || c == '_' => {}
        _ => return false,
    }
    chars.all(|c| c.is_ascii_alphanumeric() || c == '_' || c == '$')
}

/// Parses `<name> <type>[<dim>] [<distance>]`. Equivalent to the regex
/// `^(\w+)\s+(\w+)\[(\d+)\]\s*(\w+)?\s*$` used by the C++ implementation.
pub fn parse_named_vector_space(input: &str) -> Result<NamedVectorSpace, String> {
    let bytes = input.as_bytes();
    let n = bytes.len();
    let mut i = 0;

    let take_word = |start: usize| -> usize {
        let mut j = start;
        while j < n && is_word_char(bytes[j] as char) {
            j += 1;
        }
        j
    };
    let skip_ws = |start: usize| -> usize {
        let mut j = start;
        while j < n && (bytes[j] as char).is_whitespace() {
            j += 1;
        }
        j
    };

    // Leading whitespace is not allowed by the anchored regex, but the input is
    // already trimmed by SQLite's argument handling; be lenient and skip it.
    i = skip_ws(i);

    // vector name
    let name_end = take_word(i);
    if name_end == i {
        return Err("Unable to parse vector space".to_string());
    }
    let vector_name = &input[i..name_end];
    i = name_end;

    // mandatory whitespace
    let after_name_ws = skip_ws(i);
    if after_name_ws == i {
        return Err("Unable to parse vector space".to_string());
    }
    i = after_name_ws;

    // vector type
    let type_end = take_word(i);
    if type_end == i {
        return Err("Unable to parse vector space".to_string());
    }
    let vector_type_str = &input[i..type_end];
    i = type_end;

    // '['
    if i >= n || bytes[i] != b'[' {
        return Err("Unable to parse vector space".to_string());
    }
    i += 1;

    // digits
    let dim_start = i;
    while i < n && bytes[i].is_ascii_digit() {
        i += 1;
    }
    if i == dim_start {
        return Err("Unable to parse vector space".to_string());
    }
    let dim: usize = input[dim_start..i]
        .parse()
        .map_err(|_| "Unable to parse vector space".to_string())?;

    // ']'
    if i >= n || bytes[i] != b']' {
        return Err("Unable to parse vector space".to_string());
    }
    i += 1;

    // optional whitespace then optional distance word
    i = skip_ws(i);
    let mut distance_type = DistanceType::L2;
    if i < n {
        let dist_end = take_word(i);
        if dist_end == i {
            return Err("Unable to parse vector space".to_string());
        }
        let dist_str = &input[i..dist_end];
        distance_type = parse_distance_type(dist_str)
            .ok_or_else(|| format!("Invalid distance type: {dist_str}"))?;
        i = dist_end;
    }

    // only trailing whitespace permitted
    i = skip_ws(i);
    if i != n {
        return Err("Unable to parse vector space".to_string());
    }

    if !is_valid_column_name(vector_name) {
        return Err(format!("Invalid vector name: {vector_name}"));
    }

    let vector_type = parse_vector_type(vector_type_str)
        .ok_or_else(|| format!("Invalid vector type: {vector_type_str}"))?;

    Ok(NamedVectorSpace {
        vector_name: vector_name.to_string(),
        dim,
        distance_type,
        vector_type,
    })
}
